editoast/core_task/src/envs/simulation.rs

```
@@ -0,0 +1,550 @@
+pub(in crate::envs) struct SimulationKey(
+    Arc<SimulationConsist>,
+    Arc<SimulationTrainParameters>,
+    pathfinding::Input,
+);
+
+impl<Train> Runner<Train>
+where
+    Train: Clone + Hash + Eq + Send + Sync + 'static,
+{
+    pub(in crate::envs) fn new(
+        SimulationEnv {
+            pathfinding_env,
+            inputs,
+        }: SimulationEnv<Train>,
+    ) -> Self {
+        let core_env = pathfinding_env.core_env.clone();
+        let pathfinding_runner = Arc::new(pathfinding::Runner::new(pathfinding_env));
+        let rev = DashMap::<SimulationKey, TrainSet<Train>>::new();
+        for Correlated {
+            correlation_key: train,
+            data: sim_key,
+        } in inputs.iter(&pathfinding_runner.pathfinding_inputs)
+        {
+            rev.entry(
+                sim_key.expect("SimulationEnv and PathfindingEnv should have the same trains"),
+            )
+            .or_default()
+            .insert(train);
+        }
+        Self {
+            core_env,
+            simulation_inputs: Arc::new(inputs),
+            pathfinding_runner,
+            rev,
+        }
+    }
+
+    fn train_set(&self, key: &SimulationKey) -> TrainSet<Train> {
+        self.rev.get(key).expect("Runner::new invariant").clone()
+    }
+
+    /// Trains sharing the same pathfinding key (= inputs), therefore having the same path,
+    /// do not necessarily share the same simulation key (simulation inputs differ).
+    fn simulation_keys_of_pathfinding_key(
+        &self,
+        pathfinding_key: &pathfinding::Input,
+    ) -> Option<impl Iterator<Item = SimulationKey>> {
+        Some(
+            self.pathfinding_runner
+                .input_train_set(pathfinding_key)?
+                .into_iter()
+                .filter_map(|train| {
+                    self.simulation_inputs
+                        .train_input(&train, &self.pathfinding_runner.pathfinding_inputs)
+                }),
+        )
+    }
```

Isn’t this equivalent and maybe a bit more idiomatic (`Option<Iterator>` feels weird). Or is it really useful to make a distinction between the 2 cases?

```rust
    fn simulation_keys_of_pathfinding_key(
        &self,
        pathfinding_key: &pathfinding::Input,
    ) -> impl Iterator<Item = SimulationKey> {
        self.pathfinding_runner
            .input_train_set(pathfinding_key)?
            .into_iter()
            .flatten()
            .filter_map(|train| {
                self.simulation_inputs
                    .train_input(&train, &self.pathfinding_runner.pathfinding_inputs)
            }),
    )

```

It might actually be fine to also panic here instead of returning an `Option` (since this is actually what is done anyway every time this function is called).